pkg/scheduler/internal/cache/node_tree.go

```diff
@@ -135,36 +121,30 @@ func (nt *nodeTree) updateNode(old, new *v1.Node) {
 	nt.addNode(new)
 }
 
-func (nt *nodeTree) resetExhausted() {
-	for _, na := range nt.tree {
-		na.lastIndex = 0
-	}
-	nt.zoneIndex = 0
-}
-
-// next returns the name of the next node. NodeTree iterates over zones and in each zone iterates
+// list returns the list of names of the node. NodeTree iterates over zones and in each zone iterates
 // over nodes in a round robin fashion.
-func (nt *nodeTree) next() string {
+func (nt *nodeTree) list() ([]string, error) {
 	if len(nt.zones) == 0 {
-		return ""
+		return nil, nil
 	}
+	nodesList := make([]string, 0, nt.numNodes)
 	numExhaustedZones := 0
-	for {
-		if nt.zoneIndex >= len(nt.zones) {
-			nt.zoneIndex = 0
+	nodeIndex := 0
+	for len(nodesList) < nt.numNodes {
+		if numExhaustedZones >= len(nt.zones) { // all zones are exhausted.
+			return nodesList, errors.New("All zones exhausted before reaching count of nodes expected")
```

Comment:
Error string usually not start with a capitalized letter:

```suggestion
			return nodesList, errors.New("all zones exhausted before reaching count of nodes expected")
```